sdk/tables/azure-data-tables/tests/_shared/testcase.py

```diff
@@ -426,17 +440,84 @@ def build_service_endpoint(service):
+
+                if not cosmos_connection_string:
+                    # I have received a cosmos name from env
+                    cosmos_connection_string=";".join([
+                        "DefaultEndpointsProtocol=https",
+                        "AccountName={}".format(cosmos_name),
+                        "AccountKey={}".format(cosmos_key),
+                        "TableEndpoint={}".format(cosmos_account.primary_endpoints.table),
+                    ])
+
+                if not cosmos_account:
+                    cosmos_name = cosmos_connection_string_parts["AccountName"]
+
+                    def build_service_endpoint(service):
+                        try:
+                            suffix = cosmos_connection_string_parts["EndpointSuffix"]
+                        except KeyError:
+                            suffix = "cosmos.azure.com"
+                        return "{}://{}.{}.{}".format(
+                            cosmos_connection_string_parts.get("DefaultEndpointsProtocol", "https"),
+                            cosmos_connection_string_parts["AccountName"],
+                            service,
+                            suffix
+                        )
+
+                    cosmos_account.name = cosmos_name
+                    cosmos_account.id = cosmos_name
+                    cosmos_account.primary_endpoints=Endpoints()
+                    cosmos_account.primary_endpoints.table = cosmos_connection_string_parts.get("TableEndpoint", build_service_endpoint("table"))
+                    cosmos_account.secondary_endpoints=Endpoints()
+                    cosmos_account.secondary_endpoints.table = cosmos_connection_string_parts.get("TableSecondaryEndpoint", build_service_endpoint("table"))
+                    cosmos_key = cosmos_connection_string_parts["AccountKey"]
+
+            else:
+                cosmos_name, cosmos_kwargs = cosmos_preparer._prepare_create_resource(test_case, **rg_kwargs)
+                cosmos_account = cosmos_kwargs['cosmos_account']
+                cosmos_key = cosmos_kwargs['cosmos_account_key']
+                cosmos_connection_string = cosmos_kwargs['cosmos_account_cs']
+
             TableTestCase._STORAGE_ACCOUNT = storage_account
             TableTestCase._STORAGE_KEY = storage_key
             TableTestCase._STORAGE_CONNECTION_STRING = storage_connection_string
+            TableTestCase._COSMOS_ACCOUNT = cosmos_account
+            TableTestCase._COSMOS_KEY = cosmos_key
+            TableTestCase._COSMOS_CONNECTION_STRING = cosmos_connection_string
             yield
         finally:
-            if storage_name is not None:
-                if not got_storage_info_from_env:
-                    storage_preparer.remove_resource(
-                        storage_name,
-                        resource_group=rg
-                    )
+            if not got_storage_info_from_env:
+                storage_preparer.remove_resource(
```

Comment:
What if one of these fails and the other doesn't?  What if one of these didn't exist when getting into this block due to a failure earlier on?  I worry it's a recipe for dangling resources; recommend a bit more cautious internal try-catch.

Reply:
Discussed offline, opened an [issue](https://github.com/Azure/azure-sdk-for-python/issues/14184) for tracking this and removing in the next sprint.